File: packages/service-user/src/domain/model/EmailAddress.ts
export class EmailAddress {
  public readonly value: string;

  public constructor(value: string) {
    this.validateEmailAddress(value);
    this.value = value;
  }

  private validateEmailAddress(value: string) {
    if (!value) {
      throw new Error("Email address not provided");
    }
    if (
      !String(value)
        .toLowerCase()
        .match(
          /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/
        )
    ) {
      throw new Error("Email address is invalid");
    }
  }
}

File: packages/service-user/internal/core/domain/user/email_address.go
package user

import (
	"errors"
	"net/mail"
)

type EmailAddress struct {
	Value string
}

func CreateEmailAddress(value string) EmailAddress {
	err := validateEmailAddress(value)
	if err != nil {
		panic(err)
	}
	return EmailAddress{
		Value: value,
	}
}

func validateEmailAddress(value string) error {
	_, err := mail.ParseAddress(value)
	if err != nil {
		return errors.New("not a valid email address")
	}
	return nil
}
